File: Star/Star.py
# test792a : USAG-Lib star
import os
import io
import time

class TarWriter:
    def __init__(self, output: str):
        self.output = io.BytesIO() if output == "" else open(output, "wb") # set output empty to write on memory

    def _tar_header(self, name: str, size: int, mode: int, mtime: int, flag: bytes) -> bytes:
        h = bytearray(512)

        # Name (offset 0)
        name_b = name.encode('utf-8')[:100]
        h[0:len(name_b)] = name_b
        
        # Mode (100), Size (124) - oct format
        h[100:107] = f"{mode:07o}".encode()
        h[124:135] = f"{size:011o}".encode() if size < 0o77777777777 else b"00000000000"
        
        # Mtime (136) - current time
        h[136:147] = f"{mtime:011o}".encode()
        
        # Typeflag (156)
        h[156:157] = flag
        
        # Magic, Version (257)
        h[257:263] = b"ustar\x00"
        h[263:265] = b"00"
        
        # Checksum (148)
        h[148:156] = b"        "
        checksum = sum(h)
        h[148:154] = f"{checksum:06o}".encode()
        h[154:155] = b"\x00"
        return bytes(h)
    
    def _pax_header(self, name: str, size: int) -> bytes:
        pax_elements = []
        for key, value in [("path", name), ("size", str(size))]:
            line = f" {key}={value}\n" # make line
            length = len(line.encode('utf-8')) + 2
            while True: # calculate length
                full_line = f"{length}{line}"
                if len(full_line.encode('utf-8')) == length: break
                length = len(full_line.encode('utf-8'))
            pax_elements.append(full_line.encode('utf-8'))
        
        # make pax header
        pax_data = b"".join(pax_elements)
        return self._tar_header(f"PaxHeader/{name}", len(pax_data), 0o644, int(time.time()), b'x') + pax_data + self._pad(len(pax_data))

    def _pad(self, size: int) -> bytes:
        pad_size = (512 - (size % 512)) % 512
        return b'\0' * pad_size

    def WriteFile(self, name: str, path: str, mode: int = 0o644):
        size = os.path.getsize(path)
        # write pax header if long name or large size
        if len(name.encode('utf-8')) > 99 or size > 0o77777777777:
            self.output.write(self._pax_header(name, size))
        self.output.write(self._tar_header(name, size, mode, int(os.path.getmtime(path)), b'0'))

        # write file content
        with open(path, "rb") as f:
            while True:
                chunk = f.read(65536)
                if not chunk: break
                self.output.write(chunk)
        self.output.write(self._pad(size))

    def WriteDir(self, name: str, mode: int = 0o755):
        name = name.replace("\\", "/")
        if not name.endswith("/"):
            name += "/"
        if len(name.encode('utf-8')) > 99:
            self.output.write(self._pax_header(name, 0))
        self.output.write(self._tar_header(name, 0, mode, int(time.time()), b'5'))

    def WriteBin(self, name: str, data: bytes, mode: int = 0o644):
        size = len(data)
        if len(name.encode('utf-8')) > 99 or size > 0o77777777777:
            self.output.write(self._pax_header(name, size))
        self.output.write(self._tar_header(name, size, mode, int(time.time()), b'0'))
        self.output.write(data)
        self.output.write(self._pad(size))

    def Close(self) -> bytes:
        self.output.write(b"\x00" * 1024) # two 512-byte blocks of zeroes
        if isinstance(self.output, io.BytesIO):
            res = self.output.getvalue()
            self.output.close()
            self.output = None
            return res
        self.output.close()
        self.output = None
        return b""

class TarReader:
    def __init__(self, src: str|bytes):
        if isinstance(src, str):
            self.stream = open(src, "rb")
        elif isinstance(src, bytes):
            self.stream = io.BytesIO(src)
        else:
            self.stream = src

        # metadata of current entry
        self.Name = ""
        self.Size = 0
        self.Mode = 0o644
        self.IsDir = False
        self.IsEOF = False

    def _parse(self, data: bytes):
        lines = data.decode('utf-8', 'replace').split('\n')
        for line in lines: # format: "length key=value\n"
            if line == "": continue
            parts = line.split(' ', 1)
            if len(parts) < 2: continue
            kv = parts[1].split('=', 1)
            if len(kv) < 2: continue
            
            key, value = kv[0], kv[1] # update metadata by pax key-values
            if key == "path": self.Name = value
            elif key == "size": self.Size = int(value)

    def _unpad(self, size: int):
        pad = (512 - (size % 512)) % 512 # jump padding bytes
        if pad > 0: self.stream.read(pad)

    def Next(self) -> bool:
        if self.IsEOF: return False
        header = self.stream.read(512) # read next 512-byte header
        if header == None or len(header) != 512 or header == b'\0' * 512:
            self.IsEOF = True
            self.stream.read(512)
            return False

        # parse standard header
        self.Name = header[0:100].decode('utf-8', 'replace').rstrip('\0')
        self.Mode = int(header[100:108].strip(b'\x00 '), 8)
        self.Size = int(header[124:136].strip(b'\x00 '), 8)
        tp = header[156:157]
        self.IsDir = True if tp == b'5' else False

        # parse PAX header if type is 'x'
        if tp == b'x':
            pax_data = self.stream.read(self.Size)
            self._unpad(self.Size)
            self._parse(pax_data)
            tn, ts = self.Name, self.Size
            e = self.Next() # read next header after pax
            self.Name, self.Size = tn, ts
            return e
        return True

    def Read(self) -> bytes:
        data = self.stream.read(self.Size)
        self._unpad(self.Size)
        return data

    def Mkfile(self, path: str):
        if self.IsDir: # make directory
            os.makedirs(path, exist_ok=True)
            return
        with open(path, "wb") as f: # make file
            rem = self.Size
            while rem > 0:
                chunk = self.stream.read(min(rem, 65536))
                if len(chunk) == 0: break
                f.write(chunk)
                rem -= len(chunk)
        self._unpad(self.Size)

    def Skip(self):
        rem = self.Size
        while rem > 0:
            chunk = self.stream.read(min(rem, 65536))
            if len(chunk) == 0: break
            rem -= len(chunk)
        self._unpad(self.Size)

    def Close(self):
        self.stream.close()

def Pack(srcs: str | list[str], dst: str) -> None:
    if type(srcs) == str: srcs = [srcs]
    tw = TarWriter(dst)
    try:
        for src in srcs:
            if os.path.isfile(src):
                tw.WriteFile(os.path.basename(src), src)
            elif os.path.isdir(src):
                parent = os.path.dirname(os.path.normpath(src))
                for root, dirs, files in os.walk(src):
                    rel_dir = os.path.relpath(root, parent).replace("\\", "/")
                    tw.WriteDir(rel_dir)
                    for f in files:
                        path = os.path.join(root, f)
                        rel = os.path.relpath(path, parent).replace("\\", "/")
                        tw.WriteFile(rel, path)
    finally:
        tw.Close()

def Unpack(src: str, dst: str) -> None:
    tr = TarReader(src)
    try:
        while tr.Next():
            rel_path = tr.Name.replace("\\", "/")
            dest_path = os.path.join(dst, rel_path)
            
            # ZipSlip Protection
            base = os.path.join(os.path.abspath(dst), "")
            if not os.path.abspath(dest_path).startswith(base):
                raise Exception(f"illegal file path: {rel_path}")

            tr.Mkfile(dest_path)
    finally:
        tr.Close()

File: Star/Star.go
// test792c : USAG-Lib star

package Star

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"time"
)

type TarWriter struct {
	out    io.Writer
	memBuf *bytes.Buffer // Used when output is memory
	file   *os.File      // Used when output is file
}

func (tw *TarWriter) Init(output string) error {
	if output == "" { // empty string means memory
		tw.memBuf = new(bytes.Buffer)
		tw.out = tw.memBuf
	} else {
		f, err := os.Create(output)
		if err != nil {
			return err
		}
		tw.file = f
		tw.out = f
	}
	return nil
}

func (tw *TarWriter) pad(size int64) []byte {
	padSize := (512 - (size % 512)) % 512
	return make([]byte, padSize)
}

func (tw *TarWriter) tarHeader(name string, size int64, mode int, mtime int64, flag byte) []byte {
	var h [512]byte

	// Name (100)
	nameBytes := []byte(name)
	if len(nameBytes) > 100 {
		nameBytes = nameBytes[:100]
	}
	copy(h[0:], nameBytes)

	// Mode (100)
	copy(h[100:], fmt.Sprintf("%07o", mode))

	// Size (124)
	if size < 077777777777 {
		copy(h[124:], fmt.Sprintf("%011o", size))
	} else {
		copy(h[124:], "00000000000") // PAX handles real size
	}

	// Mtime (136)
	copy(h[136:], fmt.Sprintf("%011o", mtime))

	// Typeflag (156)
	h[156] = flag

	// Magic (257)
	copy(h[257:], "ustar\x00")
	copy(h[263:], "00")

	// Checksum (148) - Calc
	copy(h[148:], "        ")
	var checksum int64
	for _, b := range h {
		checksum += int64(b)
	}
	copy(h[148:], fmt.Sprintf("%06o", checksum))
	h[154] = 0
	h[155] = 32
	return h[:]
}

func (tw *TarWriter) paxHeader(name string, size int64) []byte {
	// Pax elements
	records := ""
	data := [][]string{{"path", name}, {"size", fmt.Sprintf("%d", size)}}
	for _, pair := range data {
		lineData := fmt.Sprintf(" %s=%s\n", pair[0], pair[1])
		length := len(lineData) + 1
		for { // loop until length stabilizes
			fullLine := fmt.Sprintf("%d%s", length, lineData)
			if len(fullLine) == length {
				records += fullLine
				break
			}
			length = len(fullLine)
		}
	}

	// Build PAX header block
	paxData := []byte(records)
	paxName := "PaxHeader/" + name
	header := tw.tarHeader(paxName, int64(len(paxData)), 0644, time.Now().Unix(), 'x')

	// Join: Header + Data + Pad
	buf := make([]byte, 0, 1024)
	buf = append(buf, header...)
	buf = append(buf, paxData...)
	buf = append(buf, tw.pad(int64(len(paxData)))...)
	return buf
}

func (tw *TarWriter) WriteFile(name string, path string, mode int) error {
	info, err := os.Stat(path)
	if err != nil {
		return err
	}
	size := info.Size()

	// PAX check, Write Header
	if len(name) > 99 || size > 077777777777 {
		tw.out.Write(tw.paxHeader(name, size))
	}
	tw.out.Write(tw.tarHeader(name, size, mode, info.ModTime().Unix(), '0'))

	// Write Content
	f, err := os.Open(path)
	if err != nil {
		return err
	}
	defer f.Close()
	if _, err := io.Copy(tw.out, f); err != nil {
		return err
	}
	tw.out.Write(tw.pad(size)) // Pad
	return nil
}

func (tw *TarWriter) WriteDir(name string, mode int) error {
	name = strings.ReplaceAll(name, "\\", "/")
	if !strings.HasSuffix(name, "/") {
		name += "/"
	}
	if len(name) > 99 { // PAX needed
		tw.out.Write(tw.paxHeader(name, 0))
	}
	tw.out.Write(tw.tarHeader(name, 0, mode, time.Now().Unix(), '5')) // ustar header
	return nil
}

func (tw *TarWriter) WriteBin(name string, data []byte, mode int) error {
	size := int64(len(data))
	if len(name) > 99 || size > 077777777777 {
		tw.out.Write(tw.paxHeader(name, size))
	}
	tw.out.Write(tw.tarHeader(name, size, mode, time.Now().Unix(), '0'))
	tw.out.Write(data)
	tw.out.Write(tw.pad(size))
	return nil
}

func (tw *TarWriter) Close() []byte {
	zeroes := make([]byte, 1024)
	tw.out.Write(zeroes) // write two empty blocks
	var res []byte = nil
	if tw.memBuf != nil {
		res = tw.memBuf.Bytes()
		tw.memBuf = nil
	}
	if tw.file != nil {
		tw.file.Close()
		tw.file = nil
	}
	return res
}

type TarReader struct {
	in     io.Reader
	memBuf *bytes.Buffer // Used when input is memory
	file   *os.File      // Used when input is file

	// Metadata
	Name  string
	Size  int64
	Mode  int
	IsDir bool
	IsEOF bool
}

func (tr *TarReader) Init(input interface{}) error {
	switch v := input.(type) {
	case string: // file path
		f, err := os.Open(v)
		if err != nil {
			return err
		}
		tr.file = f
		tr.in = f
	case []byte: // memory
		tr.memBuf = bytes.NewBuffer(v)
		tr.in = tr.memBuf
	default:
		return fmt.Errorf("unsupported source type")
	}

	tr.Name = ""
	tr.Size = 0
	tr.Mode = 0
	tr.IsDir = false
	tr.IsEOF = false
	return nil
}

func (tr *TarReader) unpad(size int64) {
	pad := (512 - (size % 512)) % 512
	if pad > 0 {
		io.CopyN(io.Discard, tr.in, int64(pad))
	}
}

func (tr *TarReader) parse(data []byte) {
	str := string(data)
	lines := strings.Split(str, "\n")
	for _, line := range lines {
		if line == "" {
			continue
		}
		// "length key=value"
		parts := strings.SplitN(line, " ", 2)
		if len(parts) < 2 {
			continue
		}
		kvPart := parts[1]
		kv := strings.SplitN(kvPart, "=", 2)
		if len(kv) < 2 {
			continue
		}

		key, value := kv[0], kv[1]
		if key == "path" {
			tr.Name = value
		} else if key == "size" {
			tr.Size, _ = strconv.ParseInt(value, 10, 64)
		}
	}
}

func (tr *TarReader) Next() bool {
	if tr.IsEOF {
		return false
	}
	var h [512]byte
	n, err := io.ReadFull(tr.in, h[:])
	isZero := true // Check EOF conditions
	for _, b := range h {
		if b != 0 {
			isZero = false
			break
		}
	}
	if err == io.EOF || err == io.ErrUnexpectedEOF || n != 512 || isZero {
		tr.IsEOF = true
		io.ReadFull(tr.in, h[:]) // consume next empty block if exists
		return false
	}

	// Parse Standard Header
	tr.Name = string(bytes.Trim(h[0:100], "\x00"))
	modeStr := string(bytes.Trim(h[100:108], "\x00 ")) // Trim nulls
	ti, _ := strconv.ParseInt(modeStr, 8, 64)
	tr.Mode = int(ti)
	sizeStr := string(bytes.Trim(h[124:136], "\x00 "))
	tr.Size, _ = strconv.ParseInt(sizeStr, 8, 64)
	typeFlag := h[156]
	tr.IsDir = (typeFlag == '5')

	// PAX Handling
	if typeFlag == 'x' {
		paxData := make([]byte, tr.Size)
		io.ReadFull(tr.in, paxData)
		tr.unpad(tr.Size)

		// Parse PAX
		tr.parse(paxData)
		paxName := tr.Name
		paxSize := tr.Size
		hasNext := tr.Next()
		tr.Name = paxName
		tr.Size = paxSize
		return hasNext
	}
	return true
}

func (tr *TarReader) Read() []byte {
	data := make([]byte, tr.Size)
	io.ReadFull(tr.in, data)
	tr.unpad(tr.Size)
	return data
}

func (tr *TarReader) Mkfile(path string) error {
	if tr.IsDir {
		return os.MkdirAll(path, os.FileMode(tr.Mode))
	}
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	// Copy chunked
	remaining := tr.Size
	var buf [32 * 1024]byte
	for remaining > 0 {
		toRead := int64(len(buf))
		if remaining < toRead {
			toRead = remaining
		}
		n, err := tr.in.Read(buf[:toRead])
		if n > 0 {
			f.Write(buf[:n])
			remaining -= int64(n)
		}
		if err != nil {
			break
		}
	}
	tr.unpad(tr.Size)
	return nil
}

func (tr *TarReader) Skip() {
	io.CopyN(io.Discard, tr.in, tr.Size)
	tr.unpad(tr.Size)
}

func (tr *TarReader) Close() {
	if tr.file != nil {
		tr.file.Close()
		tr.file = nil
	}
	if tr.memBuf != nil {
		tr.memBuf = nil
	}
}
